### ISSUE TITLE
Poor man's support of Nix interoperability

### ISSUE THREAD

**Issue by thufschmitt** [tweag/nickel#1431]
(Mostly a wild idea so far, testing the water to get an idea of how unreasonable it is. Also, this isn't really Nix-specific and could potentially be adapted to any language)

**Is your feature request related to a problem? Please describe.**

For https://github.com/nickel-lang/nickel-nix, having a first-class(ish) Nix interoperability would be a must. @francois-caddet started an ambitious effort to make Nickel able to evaluate Nix natively. However, there's (afaik) still a bunch of work to be done for this to fully work on Nixpkgs (which is the real goal here). So the best solution so far is a somewhat contrived and rigid one using JSON as the interop, and requiring a few back-and-forth between the different languages.

**Describe the solution you'd like**

As a first-order approximation, a Nickel unary operator `%import_nix% : String -> Dyn` that would evaluate a Nix expression (using Nix's own evaluator), and return its result as a Nickel value, replacing the functions and other non-representable values by errors.

This wouldn't be enough for a fully transparent interface, because this interface would be strict, while Nix expressions leverage heavily the laziness of the language. However, we could emulate this laziness with another primitive operator `%nix_thunk%` that takes a pointer to a (lazy) Nix value, and returns it in whnf, with all the thunks replaced by calls to `%nix_thunk%`. So for instance,

```nickel
%import_nix% "{ x = 1; y = 1+1; }"
```

would evaluate to

```nickel
{ x = 1, y = %nix_thunk% {a pointer to the Nix value representing the thunk "1+1"} } 
```

which could be further forced to

```nickel
{ x = 1, y = 2 }
```

The advantage of this approach is that the integrations is somewhat light as it “only” requires adding two builtins, meaning that it could be made optional (arguably not everyone cares about pulling in Nix when building Nickel) or even a plugin if there was such a mechanism.

**Describe alternatives you've considered**

- Fully-fledged Nix evaluator in Nickel. Cleaner (and would support functions), but bigger effort
- Keep (and try to improve) the current JSON-based FFI. I'll try to open another issue for suggestions here as we could maybe make it a bit better with some scoped changes to Nickel
- Rewrite Nixpkgs in Nickel

**Comment by yannham** [comment]
As a first step, we can guard Nix integration experimental features behind a `nix-experimental` cargo feature (that is, a compilation flag) and do quick, incremental and (not too dirty) implementation as to not block Nixel on solving the issues of "is this actually a good idea to add that to Nickel, or does it break the language". A first step with `nix_import` piggy backing on the Nix CLI wouldn't be too hard, but I wonder if that would cover any interesting use case for you, @thufschmitt? Or do even simple workflows require the thunk part?

The thunk part sounds a bit trickier. I guess it would require to link to the C++ (unstable) Nix API?

**Comment by Radvendii** [comment]
A very rudimentary version of this is implemented in #1465. If you compile `nickel` with the `nix-experimental` feature flag, you can call the `%eval_nix%` builtin with a string containing nix code.

With that PR, the nix code must fully evaluate to a json object. It can't have any thunks left over or anything like that. If that ends up being needed, let's open a new issue for it.

**Comment by thufschmitt** [comment]
Thanks @Radvendii !

Laziness would have been nice (being able to `%eval_nix% "import <nixpkgs> {}"` and then recurse through it would have been awesome :) ), but something that can be saved for later (as it's indeed making thinks quite more complex).

Something that's blocking for me with the current approach however, is that each `%eval_nix%` call re-initializes the whole Nix evaluator, meaning that (for Organist), a derivation depending on three Nix packages would have to evaluates Nixpkgs three times, which is not
